File: go/internal/tools/calculator.go
package tools

import (
	"encoding/json"
	"fmt"
)

// CalculatorInput represents the input parameters for calculator operations
type CalculatorInput struct {
	Operation string  `json:"operation"`
	A         float64 `json:"a"`
	B         float64 `json:"b"`
}

// CalculatorResult represents the result of a calculator operation
type CalculatorResult struct {
	Result    float64 `json:"result"`
	Operation string  `json:"operation"`
}

// Calculator performs basic arithmetic operations
func Calculator(args map[string]interface{}) (interface{}, error) {
	// Parse input
	argBytes, err := json.Marshal(args)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal arguments: %w", err)
	}

	var input CalculatorInput
	if err := json.Unmarshal(argBytes, &input); err != nil {
		return nil, fmt.Errorf("failed to unmarshal arguments: %w", err)
	}

	// Validate operation
	var result float64
	switch input.Operation {
	case "add":
		result = input.A + input.B
	case "subtract":
		result = input.A - input.B
	case "multiply":
		result = input.A * input.B
	case "divide":
		if input.B == 0 {
			return nil, fmt.Errorf("division by zero")
		}
		result = input.A / input.B
	default:
		return nil, fmt.Errorf("unsupported operation: %s", input.Operation)
	}

	return CalculatorResult{
		Result:    result,
		Operation: fmt.Sprintf("%g %s %g = %g", input.A, input.Operation, input.B, result),
	}, nil
}

// GetCalculatorSchema returns the JSON schema for calculator tool
func GetCalculatorSchema() map[string]interface{} {
	return map[string]interface{}{
		"type": "object",
		"properties": map[string]interface{}{
			"operation": map[string]interface{}{
				"type":        "string",
				"enum":        []string{"add", "subtract", "multiply", "divide"},
				"description": "The arithmetic operation to perform",
			},
			"a": map[string]interface{}{
				"type":        "number",
				"description": "First operand",
			},
			"b": map[string]interface{}{
				"type":        "number",
				"description": "Second operand",
			},
		},
		"required": []string{"operation", "a", "b"},
	}
}

File: rust/mcp-core/src/tools/calculator.rs
use super::ToolHandler;
use crate::types::Tool;
use anyhow::{anyhow, Result};
use serde::{Deserialize, Serialize};
use serde_json::{json, Value};
use std::collections::HashMap;

#[derive(Debug, Deserialize)]
struct CalculatorInput {
    operation: String,
    a: f64,
    b: f64,
}

#[derive(Debug, Serialize)]
struct CalculatorResult {
    result: f64,
    operation: String,
}

pub struct CalculatorTool;

#[async_trait::async_trait]
impl ToolHandler for CalculatorTool {
    fn definition(&self) -> Tool {
        Tool {
            name: "calculator".to_string(),
            description: "Perform basic arithmetic operations (add, subtract, multiply, divide)".to_string(),
            input_schema: json!({
                "type": "object",
                "properties": {
                    "operation": {
                        "type": "string",
                        "enum": ["add", "subtract", "multiply", "divide"],
                        "description": "The arithmetic operation to perform"
                    },
                    "a": {
                        "type": "number",
                        "description": "First operand"
                    },
                    "b": {
                        "type": "number",
                        "description": "Second operand"
                    }
                },
                "required": ["operation", "a", "b"]
            }),
        }
    }

    async fn execute(&self, args: HashMap<String, Value>) -> Result<Value> {
        let input: CalculatorInput = serde_json::from_value(json!(args))?;

        let result = match input.operation.as_str() {
            "add" => input.a + input.b,
            "subtract" => input.a - input.b,
            "multiply" => input.a * input.b,
            "divide" => {
                if input.b == 0.0 {
                    return Err(anyhow!("Division by zero"));
                }
                input.a / input.b
            }
            _ => return Err(anyhow!("Unsupported operation: {}", input.operation)),
        };

        Ok(json!(CalculatorResult {
            result,
            operation: format!("{} {} {} = {}", input.a, input.operation, input.b, result),
        }))
    }
}
